### DISCUSSION TITLE
Better keyboard navigation for tasks

### DISCUSSION THREAD

**Discussion by postrockreverb** [minbrowser/min#2404] [Ideas]
It would be nice if navigation through tasks was possible only by keyboard.
From the first sight it lacks of the following hot keys:
- fold/unfold all tasks
- fold/unfold task under cursor
- go next/previous task
- delete task

**Comment by PalmerAL** [comment]
I think some of this already exists:
* Next/previous task have shortcuts: `ctrl+[` and `ctrl+]`
* In the overlay, you can navigate with the tab key to the collapse and delete buttons on a task.

It would be good to have single-letter shortcuts for these though, like we have for the search bar - I'm thinking `c` for collapse, `d` for delete.

The `ctrl+[` shortcuts currently open the task and switch to it; we could modify these so that if the overlay is already open when you press them, they just switch tasks while keeping the overlay open.

How often do you collapse all tasks (and what for)? I'm wondering if this is really used often enough to add a shortcut for.

**Reply by postrockreverb** [reply]
> Next/previous task have shortcuts: ctrl+[ and ctrl+]
> In the overlay, you can navigate with the tab key to the collapse and delete buttons on a task.

To navigate between tasks in the overlay it could be `option + Up arrow` / `option + Down arrow`. I will clarify, it's only for navigation in overlay, not for actual switching tasks.

For collapsing/expanding task under cursor I would suggest `cmd + =` / `cmd + -` and for all tasks `cmd + shift + =` / `cmd + shift + -`, like in most IDEs. 

To delete task `d` or `backspace` seem good.

> The ctrl+[ shortcuts currently open the task and switch to it; we could modify these so that if the overlay is already open when you press them, they just switch tasks while keeping the overlay open.

I don't think that `ctrl+[` shortcuts should be changed, cause they stand for actual task switching and my idea is only for overlay navigation.

> How often do you collapse all tasks (and what for)? I'm wondering if this is really used often enough to add a shortcut for.

I prefer to keep all my tasks collapsed. This way I can easily find exact one task, especially if I have too many tabs opened. But when I need to find a tab, I'll have to expand all the tasks and then collapse them back until I find.

Or it may be better to support userscript `@run-at tasks-overlay`, but seems like a big overhead.

**Reply by PalmerAL** [reply]
> To navigate between tasks in the overlay it could be option + Up arrow / option + Down arrow

That sounds good to me.

> For collapsing/expanding task under cursor I would suggest cmd + = / cmd + - and for all tasks cmd + shift + = / cmd + shift + -, like in most IDEs.

We already use these for zooming a page in and out; I think it would be confusing to have them do something different while the overlay is open.

> Or it may be better to support userscript `@run-at tasks-overlay`, but seems like a big overhead.

Userscripts only run inside the context of a webpage, so they don't have access to anything in the browser UI (and if you let userscripts run there, they would have full access to your computer, which is fairly bad from a security perspective. To do this correctly, we'd need to implement an extension permission model like Chrome has, which is a lot of work).

**Reply by postrockreverb** [reply]
> > For collapsing/expanding task under cursor I would suggest cmd + = / cmd + - and for all tasks cmd + shift + = / cmd + shift + -, like in most IDEs.

> We already use these for zooming a page in and out; I think it would be confusing to have them do something different while the overlay is open.

I can also suggest `c` / `e` for collapse / expand.

Is it possible to prevent zooming a page (or even handling any other keys on the page) while the overlay is opened?
I believe some users, including me, may wish to reassign these to zooming keys :)
